[EDCC] Add new object type for EDCC

KUMARMUKULSAP opened the pull request.

GuilhermeSaraiva96 reviewed (changes requested):
Hi, thank you for your aff. Definitely one of the biggest I have seen :)

KUMARMUKULSAP commented:
@GuilhermeSaraiva96 : I have incorporated the above suggested changes. Please have a look and suggest if we can have a meeting to understand this AFF.

KUMARMUKULSAP commented:
@GuilhermeSaraiva96 : Let us know if the AFF requires further adjustment

GuilhermeSaraiva96 commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,600 @@
+    ty_check_formula        TYPE c LENGTH 1.
+  TYPES:
+    "! <p class="shorttext">Comparison Field Name</p>
+    "! Field Names those values will be compared
```
```suggestion
    "! Field Names whose values will be compared
```

GuilhermeSaraiva96 commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,600 @@
+    END OF ty_comparison_field.
+  TYPES:
+    "! <p class="shorttext">Comparison Field Name</p>
+    "! Field Names those values will be compared
```
```suggestion
    "! Field Names whose values will be compared
```

GuilhermeSaraiva96 commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,600 @@
+      "! Level of comparison field group
+      field_group_level TYPE ty_field_group_level,
+      "! <p class="shorttext">Check Formula</p>
+      "! Formula applied to perform the content check on field group
```
```suggestion
      "! Formula applied to perform the content check on the field group
```

GuilhermeSaraiva96 commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,600 @@
+      "! Result process
+      result_process    TYPE ty_result_process,
+      "! <p class="shorttext">Comparison Field Name</p>
+      "! Field Names those values will be compared
```
```suggestion
      "! Field Names whose values will be compared
```

GuilhermeSaraiva96 commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,600 @@
+      sequence_number    TYPE ty_sequence_number,
+      "! <p class="shorttext">Source Field Name 1</p>
+      "! Source field name 1
+      source_fieldname1  TYPE zif_aff_types_v1=>ty_object_name_30,
```
```suggestion
      source_field_name_1  TYPE zif_aff_types_v1=>ty_object_name_30,
```

GuilhermeSaraiva96 replied:
the others below as well. It is more readable

GuilhermeSaraiva96 replied:
missing

GuilhermeSaraiva96 commented:
Hi, thank you for your changes, looks good already. I have just a few more suggestions. Additionally, could you please send a meeting request so that we can discuss the whole AFF before moving to the UX review? Thanks!

GuilhermeSaraiva96 commented on file-formats/edcc/examples/z_aff_example_edcc.edcc.json:
```diff
@@ -0,0 +1,136 @@
+          "checkId": "EDCSTAT",
+          "description": "Example Status Check",
+          "sequenceNumber": "1",
+          "sourceFieldname1": "InvoiceReportingStatus",
```
this example json needs to be generated again. These properties still have the old names

KUMARMUKULSAP commented:
@GuilhermeSaraiva96 : I have regenerated the examples and incorporated changes related to description of objects coming from other object types like EDCR/EDCK. Additionally I have added one new node in the AFF in ty_main,  additional_selection_fields. Please have a look and let me know.

GuilhermeSaraiva96 commented:
From my side, it is ready for the UX-Review. @Nicolas?
Note for the UX Colleagues: the country field is an attribute directly at the root level. Should it be wrapped into a 'General Information' structure? Then it will be shown in the selection Detail instead of as grid in the root level

GuilhermeSaraiva96 commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,630 @@
+    ty_short_description TYPE c LENGTH 30.
+  TYPES:
+    "! <p class="shorttext">Tax Authority Message Type</p>
+    "! Tax Authority Message type
```
```suggestion
    "! Tax authority message type
```

GuilhermeSaraiva96 commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.json:
```diff
@@ -0,0 +1,7 @@
+{
+  "formatVersion": "1",
+  "header": {
+    "description": "AFF type for EDCC",
```
would you like to add in () what EDCC stands for?

KUMARMUKULSAP commented:
@GuilhermeSaraiva96 : Above comments are fixed. please have a look

GuilhermeSaraiva96 commented:
ready for UX Review

huber-nicolas commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,588 @@
+      "! <p class="shorttext">Country</p>
+      "! Country
+      "! $required
+      country                     TYPE c LENGTH 3,
```
Can you please put the country field into a structure general_information

schneidermic0 replied:
This is still open, isn't it?

KUMARMUKULSAP replied:
@schneidermic0 This was initially pointed out by Guilherme and it was then later decided to not put it into general information rather put it as separate field. I cannot recall the reason but @GuilhermeSaraiva96 including you into this discussion to share your input, if you can recall.

schneidermic0 replied:
I would be also interested in the reason for this decision. Here is our general rule documented (we have added some years ago):

> Top level fields shall be collected in one sub structure (ideally with a specific name) or within the structure general_information.

See https://github.com/SAP/abap-file-formats/blob/main/docs/json.md#writing-json-schema-with-abap-types

KUMARMUKULSAP replied:
@schneidermic0 : I found the comment by Guilherme around this discussion. 
https://github.com/SAP/abap-file-formats/pull/661#issuecomment-2425886813
Please let me know if it is ok to put country information in General Information.

wurzka replied:
Please put the field "country" in a substructure called "general_information" to follow our guidelines

KUMARMUKULSAP replied:
Adjusted

wurzka commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,588 @@
+      check_id           TYPE zif_aff_types_v1=>ty_object_name_30,
+      "! <p class="shorttext">Position Number</p>
+      "! Position number
+      position_number    TYPE ty_check_position_number,
```
Again, number is character like which might be confusing.

JyotishkaS replied:
@wurzka  - position_number is a numberic position number, which other data type we can use ?
cc @KUMARMUKULSAP

wurzka replied:
As mentioned above, n length 2 would be possible to restrict to numbers

KUMARMUKULSAP replied:
@wurzka 
This field has been commented out

wurzka commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,588 @@
+  TYPES:
+    "! <p class="shorttext">Selection Filters on Check</p>
+    "! Selection filters on check
+    BEGIN OF ty_check_filter,
```
Anything required?

wurzka commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,588 @@
+      "! <p class="shorttext">Check Execution Control</p>
+      "! Check execution control
+      execution_control  TYPE ty_execution_control,
+      "! <p class="shorttext">Selection Filters on Check</p>
```
De we really need `On Check`? We edit Check Attributes, therefore I don't think that it is necessary
```suggestion
      "! <p class="shorttext">Selection Filters</p>
```

JyotishkaS replied:
@wurzka  - We wanted to convey this fact that there could be few filters configured which will be considered during the run of a consistency check run. Does it make sense to have only "_Selection Filters_" or "_Selection Filters on Check_" ?

wurzka replied:
In this case, ""Selection Filters on Check" is fine for me.
You can think about renaming "check_filter" then to "filters_on_check" or "selection_filters_on_check" to emphasize this also in the JSON data file.

KUMARMUKULSAP replied:
@wurzka : This section has been commented out. As we will plan to ship this node in forthcoming releases

wurzka commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,588 @@
+      "! Expression indicator
+      expression_indicator TYPE ty_expression_indicator,
+    END OF ty_check_filter.
+  "! <p class="shorttext">Selection Filters on Check</p>
```
```suggestion
  "! <p class="shorttext">Selection Filters</p>
```

wurzka commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,588 @@
+  "! $default {@link zif_aff_edcc_v1.data:co_expression.begin_of}
+  TYPES ty_expression_indicator TYPE c LENGTH 3.
+  TYPES:
+    "! <p class="shorttext">Selection Filters on Check</p>
```
```suggestion
    "! <p class="shorttext">Selection Filters</p>
```

wurzka commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,588 @@
+      "! <p class="shorttext">Comparison Operator</p>
+      "! Comparison operator
+      comparison_operator  TYPE ty_comparison_operator,
+      "! <p class="shorttext">Filter Field Value</p>
```
Do we need field?
```suggestion
      "! <p class="shorttext">Filter Value</p>
```

JyotishkaS replied:
@wurzka  - we will incorporate this change.
cc. @KUMARMUKULSAP

wurzka commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,588 @@
+      "! <p class="shorttext">Representation Types</p>
+      "! Assigned representation types
+      "! $required
+      representation_types TYPE ty_representation_types,
```
These Representation Types have been already part in Content Checks Comparism Field Groups. Is there a reason why they are here twice?

JyotishkaS replied:
@wurzka  - Yes, these representation types are there as part of content checks comparison in order to define a combination of which field from the respective representation type will be considered as part of content check. However, the current node representation types denote the ones which developer has to define against a consistency scenario. Please let me know if any changes required.

wurzka replied:
So can the two lists be different? Is it possible that in Comparison Type-Representation Types are different values than in the top level comparison type table? And the value of the Representation Type in a Comparison Field Name is different from any value that is defined in the two other tables?

JyotishkaS replied:
@wurzka -No the lists cannot be different. The representation types onboarded in the step "Relationship Attributes" could only be used in the individual comparison type representation types of node. In fact, the content check comparison groups come after the relationship attributes.

wurzka commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,588 @@
+    "! <p class="shorttext">Representation Type</p>
+    "! Assigned representation type
+    BEGIN OF ty_representation_type,
+      "! <p class="shorttext">Assigned Representation Type</p>
+      "! Assigned representation type
+      representation_type TYPE zif_aff_types_v1=>ty_object_name_30,
```
This 
```suggestion
      "! <p class="shorttext">Type</p>
      "! Assigned representation type
      type TYPE zif_aff_types_v1=>ty_object_name_30,
```
or that 
```suggestion
      "! <p class="shorttext">Representation Type</p>
      "! Assigned representation type
      representation_type TYPE zif_aff_types_v1=>ty_object_name_30,
```
should be enough

wurzka commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,588 @@
+  TYPES:
+    "! <p class="shorttext">Check Attribute</p>
+    "! Check attribute
+    BEGIN OF ty_check_attribute,
```
There seems to be check attributes for the checks defined before. Is there a reason why you put the check attributes as a separate structure and not directly to the checks? For example as a sub structure?

JyotishkaS replied:
@wurzka  - Check Attributes is the first node which properties of a check ID is defined. Place where the position, execution sequence, filters will be defined. This configuration is irrespective of which comparison types the check IDs are assigned. They are check dependent.
Please let me know if you need any changes here.

wurzka replied:
So every check that you can define have the same structure of Check Attribute.
So, when you have an existence_check with check_id = 1, you define for this check in check_attributes an item with check_id = 1.
So check_id is the identifier for the checks and you can refer to checks defined in existence_check, status_check and content_check via this identifier. Is my understanding correct?

wurzka replied:
What do you think about adding the fields `Check Execution Sequence` and `Check Execution Control` to the dedicated checks? Or is it possible to have a check that has no entry in this table?

wurzka commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,588 @@
+      "! Define comparison types
+      "! $required
+      comparison_types            TYPE ty_comparison_types,
+      "! <p class="shorttext">Events Assignment</p>
```
Should be enough:
```suggestion
      "! <p class="shorttext">Events</p>
```

JyotishkaS replied:
@wurzka - We will incorporate the changes.
cc @KUMARMUKULSAP

wurzka commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,588 @@
+      "! Define relevance of checks
+      check_relevance     TYPE ty_check_relevances,
+    END OF ty_event.
+  "! <p class="shorttext">Assign Events</p>
```
```suggestion
  "! <p class="shorttext">Events</p>
```

JyotishkaS replied:
@wurzka  - We will incorporate the changes.
cc @KUMARMUKULSAP

wurzka commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,588 @@
+  TYPES ty_check_relevances TYPE SORTED TABLE OF ty_check_relevance WITH UNIQUE KEY
+                              comparison_type check_id.
+  TYPES:
+    "! <p class="shorttext">Assign Event</p>
```
```suggestion
    "! <p class="shorttext">Event</p>
```

JyotishkaS replied:
@wurzka  - We will incorporate the changes.
cc @KUMARMUKULSAP

wurzka commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,588 @@
+    "! <p class="shorttext">Assign Event</p>
+    "! Assign event to consistency scenario
+    BEGIN OF ty_event,
+      "! <p class="shorttext">Event Name</p>
+      "! Event name
+      event               TYPE ty_event_name,
```
```suggestion
      "! <p class="shorttext">Name</p>
      "! name
      name               TYPE ty_event_name,
```

wurzka commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,588 @@
+      "! <p class="shorttext">Event Name</p>
+      "! Event name
+      event               TYPE ty_event_name,
+      "! <p class="shorttext">Assign Checks and Comparison Type</p>
+      "! Assign Checks and Comparison Type
+      assigned_comparison TYPE ty_assigned_comparisons,
```
```suggestion
      "! <p class="shorttext">Checks and Comparison Types</p>
      "! Assign Checks and Comparison Type
      checks_and_comparison_types TYPE ty_assigned_comparisons,
```

wurzka replied:
Besides that, this seems to be again kind of double information. Checks and Comparison Types have been there before. 
Each check belongs to a comparison type. Why do we need this here again?

JyotishkaS replied:
@wurzka - This _assigned checks and comparison types_ are specific to an "Event" assigned to a consistency scenario. The "event" is an optional configuration for developers. Once the "event" is configured for a consistency scenario, developers need to take care of the checks and their corresponding comparison types which they want to run as part of the "event". They will get to choose only the checks and comparison types which are configured much earlier to the consistency scenario. 
Please let know if this needs any change.

wurzka replied:
So you define an event and assign a comparison type from before in this array. In this table, there are two field, comparison type and check id. Do you always have to add the checks of a comparison type as well?
And for Check Relevance, can you also configure the relevance for Comparison Type and Check ID added in the assigned checks and comparison types?

wurzka commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,588 @@
+      "! <p class="shorttext">Events Assignment</p>
+      "! Assign events to consistency scenario
+      events                      TYPE ty_events,
+      "! <p class="shorttext">Representation Types</p>
+      "! Assign representation types associated with consistency scenario
+      "! $required
+      relationship_attribute      TYPE ty_relationship_attributes,
```
Title and name of component shall be in sync.  
However, there is already a subcomponent Representation Types in Content Checks. Do we need them twice?

wurzka commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,588 @@
+    BEGIN OF ty_relationship_attribute,
+      "! <p class="shorttext">Representation Type</p>
+      "! Representation type
+      representation_type TYPE zif_aff_types_v1=>ty_object_name_30,
```
Required?

wurzka commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,588 @@
+      "! Assign representation types associated with consistency scenario
+      "! $required
+      relationship_attribute      TYPE ty_relationship_attributes,
+      "! <p class="shorttext">eDocument Type Assignment</p>
+      "! Assign eDocument types associated with consistency scenario
+      edocument_types             TYPE ty_edoc_types,
```
```suggestion
      "! <p class="shorttext">eDocument Types</p>
      "! Assign eDocument types associated with consistency scenario
      edocument_types             TYPE ty_edoc_types,
```

JyotishkaS replied:
@wurzka  - We will incorporate these changes.
cc @KUMARMUKULSAP

wurzka commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,588 @@
+      "! eDocument Type
+      edoc_type TYPE c LENGTH 10,
+    END OF ty_edoc_type.
+  "! <p class="shorttext">eDocument Type Assignment</p>
```
```suggestion
  "! <p class="shorttext">eDocument Types</p>
```

JyotishkaS replied:
@wurzka  - We will incorporate these changes.
cc. @KUMARMUKULSAP

wurzka commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,588 @@
+    "! <p class="shorttext">eDocument Type Assignment</p>
+    "! Assign eDocument types associated with consistency scenario
+    BEGIN OF ty_edoc_type,
+      "! <p class="shorttext">eDocument Type</p>
+      "! eDocument Type
+      edoc_type TYPE c LENGTH 10,
```
Or only 
```suggestion
      "! <p class="shorttext">Type</p>
      "! eDocument Type
      type TYPE c LENGTH 10,
```

wurzka replied:
Should this be a mandatory field?

JyotishkaS replied:
@wurzka  - No this is not mandatory.

wurzka replied:
So 
"edocumentTypes": [ {} ] 
is valid from your site?

KUMARMUKULSAP replied:
added required tag

wurzka commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,588 @@
+    "! <p class="shorttext">Inconsistency Category</p>
+    "! Inconsistency category
+    BEGIN OF ty_inconsistency_category,
+      "! <p class="shorttext">Inconsistency Category</p>
+      "! Inconsistency category
+      result_ui_group  TYPE ty_resultgroup,
```
```suggestion
      "! <p class="shorttext">Category</p>
      "! Inconsistency category
      category  TYPE ty_resultgroup,
```

JyotishkaS replied:
@wurzka  - We will incorporate this change.
cc @KUMARMUKULSAP

wurzka commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,588 @@
+      country_xtension TYPE zif_aff_types_v1=>ty_object_name_30,
+      "! <p class="shorttext">Result Processes</p>
+      "! Assign result process to the UI group
+      result_process   TYPE ty_result_processes,
```
These result processes have been there before (field of checks). Do we need them twice?

JyotishkaS replied:
@wurzka - Yes, we need the configuration of result process twice (at comparison type-->existence check/status check/content check level, and at inconsistency category level) since the use of both configurations are separate.

wurzka replied:
ok, is fine for me

wurzka reviewed (changes requested):
Please see my comments

JyotishkaS commented:
Hi @wurzka and @GuilhermeSaraiva96  - Following are the justifications for the queries asked:
1. The table name is not mandatory, and it's intended to configure a structure/table in it. However, if a table/structure name is configured there, then the field should be mandatory, which we intend to handle during side-effect handling class development of this field.
2. Since there can be more than 1 comparison types assigned to a consistency scenario, hence, to categorize all settings of a single comparison type, we have named the individual node as comparison type and not just "type". This is not a separate ABAP object and defined in EDCC alongside consistency scenarios.
3. eDocument types - This is also not mandatory for consistency scenario.
4. Yes, we need the configuration of result process twice (at comparison type-->existence check/status check/content check level, and at inconsistency category level) since the use of both configurations are separate.


Do you still see to rename the comparison type to "Type" and Inconsistency Category to "Category"?
cc @KUMARMUKULSAP

wurzka commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,588 @@
+      "! <p class="shorttext">Country</p>
+      "! Country
+      "! $required
+      country                     TYPE c LENGTH 3,
```
Please also answer to this question if possible.

JyotishkaS replied:
@wurzka this is country, not language. Do you see this as an issue ?

KUMARMUKULSAP replied:
@wurzka : At design time, consistency scenarios are created for specific countries. For example: we would develop consistency scenario for checking customer invoices for Spain but not for other countries. There is a licensing check that is performed on customising views to provide access to consistency scenarios for the country that is enabled for the SAP user.

wurzka replied:
thanks for the explanation.

JyotishkaS commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,588 @@
+    "! <p class="shorttext">Additional Selection Field</p>
+    "! Additional selection fields of validation report
+    BEGIN OF ty_additional_selection_field,
+      "! <p class="shorttext">Table Name</p>
+      "! Table name
+      table_name TYPE zif_aff_types_v1=>ty_object_name_30,
+      "! <p class="shorttext">Field Name</p>
+      "! Field name
+      field_name TYPE zif_aff_types_v1=>ty_object_name_30,
+      "! <p class="shorttext">Field Type</p>
+      "! Field type
+      field_type TYPE ty_field_type,
```
@wurzka  - The table name is not mandatory, and it's intended to configure a structure/table in it. However, if a table/structure name is configured there, then the field should be mandatory, which we intend to handle during side-effect handling class development of this field.

wurzka replied:
`ty_additional_selection_field` is the line type of `ty_additional_selection_fields`, i.e. the component "additionalSelectionFields" in the JSON. Is this possible from your point of you?
"additionalSelectionFields": [ {} ]
This means having an item in the table that is completely empty? From you specification, this is the case.
But if any entry in the array/table requires a "tableName", "fieldName" or whatever, at least one of these components should be required.
Having such a component required does not mean that the table/array must have at least one entry. It can be always empty, i.e., "additionalSelectionFields": [  ]

KUMARMUKULSAP replied:
I have added "! $required to the definition of individual field. As the node itself is mandatory, but if the value of the node is filled, then a valid entry must have all 3 values filled. Next push would have this change

JyotishkaS commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,588 @@
+    "! <p class="shorttext">Additional Tax Authority Table</p>
+    "! Additional tax authority table
+    BEGIN OF ty_tax_authority_table,
+      "! <p class="shorttext">Table Name</p>
+      "! Table name
+      table_name TYPE zif_aff_types_v1=>ty_object_name_30,
```
@wurzka  - The table name is not mandatory, and it's intended to configure a structure/table in it. However, if a table/structure name is configured there, then the field should be mandatory, which we intend to handle during side-effect handling class development of this field.

JyotishkaS commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,588 @@
+    "! <p class="shorttext">Comparison Type</p>
+    "! Comparison type
+    BEGIN OF ty_comparison_type,
+      "! <p class="shorttext">Comparison Type</p>
+      "! Comparison type
+      "! $required
+      comparison_type      TYPE zif_aff_types_v1=>ty_object_name_30,
```
@wurzka  - Since there can be more than 1 comparison types assigned to a consistency scenario, hence, to categorize all settings of a single comparison type, we have named the individual node as comparison type and not just "type". This is not a separate ABAP object and defined in EDCC alongside consistency scenarios.

wurzka replied:
This way, we have duplicated information:
"comparisonTypes": [
    {
      "comparisonType": "TaxAuthV/sEdoc",
      "description": "Tax Authority V/s Edocument",

Hence, the term "comparisonType" is somehow duplicated. Changing its name to "name" or "typeName" could stop us from repeating ourselves, so the example would be
"comparisonTypes": [
    {
      "name": "TaxAuthV/sEdoc",
      "description": "Tax Authority V/s Edocument",

With that, you can still have more than 1 comparison type.

JyotishkaS commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,588 @@
+      "! Comparison type
+      "! $required
+      comparison_type      TYPE zif_aff_types_v1=>ty_object_name_30,
+      "! <p class="shorttext">Comparison Type Description</p>
+      "! Description of the comparison type
+      "! $required
+      description          TYPE ty_long_description,
```
@wurzka  - Since there can be more than 1 comparison types assigned to a consistency scenario, hence, to categorize all settings of a single comparison type, we have named the individual node as comparison type and not just "type". This is not a separate ABAP object and defined in EDCC alongside consistency scenarios.

wurzka replied:
Same as https://github.com/SAP/abap-file-formats/pull/661#discussion_r1959627789

JyotishkaS commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,588 @@
+      "! <p class="shorttext">Position Number</p>
+      "! Position number
+      "! $required
+      position_number      TYPE ty_position_number,
```
@wurzka -No this will always be a number having length of 2.

wurzka replied:
Wouldn't then be the correct ABAP type be n length 2?

JyotishkaS commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,588 @@
+    BEGIN OF ty_existence_check,
+      "! <p class="shorttext">Check Id</p>
+      "! Check id
+      check_id       TYPE zif_aff_types_v1=>ty_object_name_30,
```
@wurzka  - Check ID is a common term as this would internal call the check IDs configured in EDCK which is an ABAP object.

wurzka replied:
ok, then it's fine for me

JyotishkaS commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,588 @@
+  TYPES:
+    "! <p class="shorttext">Existence Check</p>
+    "! Assigned existence check
+    BEGIN OF ty_existence_check,
```
@wurzka - check ID, field name, field value and result process is already within TY_EXISTENCE_CHECK.

wurzka replied:
Yeah, but shouldn't at least one of these fields be a mandatory field?
So far, only existence_check is required. But then it's possible having
"existenceCheck": {}
No checkId, FieldName,.. needs to be specified.

JyotishkaS commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,588 @@
+      "! Position number
+      "! $required
+      position_number      TYPE ty_position_number,
+      "! <p class="shorttext">Implementing Class</p>
+      "! Implementing Class
+      "! $required
+      class_name           TYPE zif_aff_types_v1=>ty_object_name_30,
```
@wurzka  - We will incorporate the same in our AFF.
cc @KUMARMUKULSAP

JyotishkaS commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,588 @@
+      "! <p class="shorttext">Result Process</p>
+      "! Result process
+      result_process    TYPE ty_result_process,
+      "! <p class="shorttext">Comparison Field Name</p>
+      "! Field Names whose values will be compared
+      comparison_field  TYPE ty_comparison_fields,
```
@wurzka  - We will incorporate the same in our AFF.
cc @KUMARMUKULSAP

JyotishkaS commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,588 @@
+      "! Comparison sequence
+      comparison_sequence TYPE ty_comparison_sequence,
+    END OF ty_comparison_field.
+  "! <p class="shorttext">Comparison Field Name</p>
```
@wurzka - Does it makes sense to have "_Field_" instead of "_Fields_" since at a time only single field can be configured in a single TY_COMPARISON_FIELD?

wurzka replied:
No, in this case I refer to the title and description of ty_comparison_fields, i.e. a table with (possibly) multiple fields

JyotishkaS commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,588 @@
+  "! $default {@link zif_aff_edcc_v1.data:co_check_formula.field}
+  TYPES ty_check_formula       TYPE c LENGTH 1.
+  TYPES:
+    "! <p class="shorttext">Comparison Field Name</p>
+    "! Field Names whose values will be compared
```
@wurzka  - we will incorporate this change.
cc @KUMARMUKULSAP

wurzka commented on file-formats/edcc/examples/z_aff_example_edcc.edcc.json:
```diff
@@ -0,0 +1,144 @@
+        "fieldValue": "X",
+        "resultProcess": "EXMISMATCH"
+      },
+      "statusCheck": [
```
This is an array, does it make sense to have it in plural?
```suggestion
      "statusChecks": [
```

wurzka reviewed (changes requested):
Thanks for your answers. I'm still working on some of them, but feel free to react to the questions and comments I already posted.

KUMARMUKULSAP commented:
@wurzka : I have committed the latest changes. However, I am unable to understand the automated error for example class. Can you please help? All the comments those have been resolved in the current AFF has been marked as resolved, leaving the set of questions where I have provided my answers.

wurzka commented:
@KUMARMUKULSAP Thanks for updating the AFF. We will have a look into it.
In the example, some required fields like "formatVersion" and "description" in the "header" are missing. I checked the previous example. Could you add
`"formatVersion": "1",
  "header": {
    "description": "Sample Consistency Scenario",
    "originalLanguage": "en",
    "abapLanguageVersion": "cloudDevelopment"
  },`
to the example?

KUMARMUKULSAP commented:
> @KUMARMUKULSAP Thanks for updating the AFF. We will have a look into it. In the example, some required fields like "formatVersion" and "description" in the "header" are missing. I checked the previous example. Could you add
> 
> ```json
> "formatVersion": "1",
>   "header": {
>     "description": "Sample Consistency Scenario",
>     "originalLanguage": "en",
>     "abapLanguageVersion": "cloudDevelopment"
>   },
> ```
> 
> to the example?

Adjusted the changes as per suggestion and it worked. Thank you. Reuested for a re-review

wurzka commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,608 @@
+      "! Sequence number
+      "! $required
+      sequence_number      TYPE ty_sequence_number,
+      "! <p class="shorttext">Source Field Name 1</p>
+      "! Source field name 1
+      "! $required
+      source_field_name_1  TYPE zif_aff_types_v1=>ty_object_name_30,
```
For the existence check, you only have `field_name` and `field_value`. Is the term `source` important for the status check? Furthermore, I wondered wether 
```suggestion
      "! <p class="shorttext">First Source Field Name</p>
      "! First source field name
      "! $required
      first_source_field_name  TYPE zif_aff_types_v1=>ty_object_name_30,
```
(or without `source`) is better as numbers in the names are quite unusal.

wurzka commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,608 @@
+    "! <p class="shorttext">Existence Check</p>
+    "! Assigned existence check
+    BEGIN OF ty_existence_check,
+      "! <p class="shorttext">Check Id</p>
```
```suggestion
      "! <p class="shorttext">Check ID</p>
```

wurzka commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,608 @@
+    "! <p class="shorttext">Status Check</p>
+    "! Status check
+    BEGIN OF ty_status_check,
+      "! <p class="shorttext">Check Id</p>
```
```suggestion
      "! <p class="shorttext">Check ID</p>
```

wurzka commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,608 @@
+    "! <p class="shorttext">Content Check</p>
+    "! Content check
+    BEGIN OF ty_content_check,
+      "! <p class="shorttext">Check Id</p>
```
```suggestion
      "! <p class="shorttext">Check ID</p>
```

wurzka commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,608 @@
+      "! <p class="shorttext">Comparison Field Groups</p>
+      "! Comparison field groups
+      "! $required
+      content_comparison_field_group TYPE ty_comparison_field_groups,
```
```suggestion
      comparison_field_groups TYPE ty_comparison_field_groups,
```

wurzka commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,608 @@
+    "! <p class="shorttext">Comparison Field Group</p>
+    "! Comparison field group
+    BEGIN OF ty_comparison_field_group,
+      "! <p class="shorttext">Mismatch Field Group</p>
+      "! Comparison field group for mismatched fields
+      "! $required
+      field_group       TYPE ty_field_group,
```
If `Mismatch` is important, I would also add it to the name of the component
```suggestion
      "! <p class="shorttext">Mismatch Field Group</p>
      "! Comparison field group for mismatched fields
      "! $required
      mismatch_field_group       TYPE ty_field_group,
```
If not, remove it. In this case, it might even make sense to change it to 
```suggestion
      "! <p class="shorttext">Name</p>
      "! Comparison field group for mismatched fields
      "! $required
      name       TYPE ty_field_group,
```

KUMARMUKULSAP replied:
mismatch is important information. I have kept it and added it to the component name. Adjusted the type name accordingly

wurzka commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,608 @@
+      "! Mismatch field group description
+      "! $required
+      description       TYPE ty_short_description,
+      "! <p class="shorttext">Field Group Level</p>
+      "! Level of comparison field group
+      "! $required
+      field_group_level TYPE ty_field_group_level,
```
Would only `level` be enough?
```suggestion
      "! <p class="shorttext">Level</p>
      "! Level of comparison field group
      "! $required
      level TYPE ty_field_group_level,
```

wurzka commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,608 @@
+      "! Result process
+      "! $required
+      result_process    TYPE ty_result_process,
+      "! <p class="shorttext">Comparison Field Name</p>
+      "! Field Names whose values will be compared
+      "! $required
+      comparison_field  TYPE ty_comparison_fields,
```
What do you think about the following?
```suggestion
      "! <p class="shorttext">Comparison Fields</p>
      "! Field Names whose values will be compared
      "! $required
      comparison_fields  TYPE ty_comparison_fields,
```

wurzka commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,608 @@
+      "! Comparison field groups
+      "! $required
+      content_comparison_field_group TYPE ty_comparison_field_groups,
+*      "! <p class="shorttext">Selection Filters on Check</p>
+*      "! Selection filters on check
+*      check_filter       TYPE ty_check_filters,
```
Please remove it if it is not needed.

wurzka commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,608 @@
+      "! Define comparison types
+      "! $required
+      comparison_types         TYPE ty_comparison_types,
+      "! <p class="shorttext">Events Assignment</p>
+      "! Assign events to consistency scenario
+      events                   TYPE ty_events,
```
```suggestion
      "! <p class="shorttext">Events</p>
      "! Assign events to consistency scenario
      events                   TYPE ty_events,
```

KUMARMUKULSAP replied:
To be resolved after gaining more clarity via below question[https://github.com/SAP/abap-file-formats/pull/661#discussion_r2024109724](url)

wurzka commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,608 @@
+    "! <p class="shorttext">Assign Event</p>
+    "! Assign event to consistency scenario
+    BEGIN OF ty_event,
+      "! <p class="shorttext">Event Name</p>
+      "! Event name
+      "! $required
+      event               TYPE ty_event_name,
```
As the name of the table is already `Events`, `Name` might be enough here.
```suggestion
      "! <p class="shorttext">Name</p>
      "! Event name
      "! $required
      name               TYPE ty_event_name,
```

wurzka commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,608 @@
+      "! Event name
+      "! $required
+      event               TYPE ty_event_name,
+      "! <p class="shorttext">Assign Checks and Comparison Type</p>
+      "! Assign Checks and Comparison Type
+      "! $required
+      assigned_comparison TYPE ty_assigned_comparisons,
```
Do we need the `Assign`?
```suggestion
      "! <p class="shorttext">Checks and Comparison Type</p>
      "! Assign Checks and Comparison Type
      "! $required
      assigned_comparison TYPE ty_assigned_comparisons,
```
If you want to keep it, I'd change it to
```suggestion
      "! <p class="shorttext">Assigned Checks and Comparison Type</p>
      "! Assign Checks and Comparison Type
      "! $required
      assigned_comparison TYPE ty_assigned_comparisons,
```

KUMARMUKULSAP replied:
@wurzka : I used this example, as I wanted to highlight the activity that the customer is supposed to do. I need clarity on the names that should appear on the UI. Should it be with the pattern <Object + Activity> or <Activity + Object>.  For example
1. Edocument type assignment to the scenario is represented as 'eDocument Type Assignment' in UI
2. while check assignment to comparison type is represented as 'Assign Checks' in UI. 

Can you please help here?
<img width="1129" alt="Screenshot 2025-04-02 at 11 16 01 AM" src="https://github.com/user-attachments/assets/83ce80de-6251-48ad-ae83-aa0d1f333b4d" />

wurzka replied:
Usually, we do not use labels that (literally) tell the user what he has to do. Therefore, I suggest not using `Assign....`.
I'm wondering whether we need the activity in the names because adding an eDocument Type or a Check is somehow an assignment. Therefore, `eDocument Types` and `Checks` might be even enough.
@schneidermic0 what is your opinion on that?

wurzka commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,608 @@
+      "! Comparison type
+      "! $required
+      comparison_type TYPE zif_aff_types_v1=>ty_object_name_30,
+      "! <p class="shorttext">Check Id</p>
```
```suggestion
      "! <p class="shorttext">Check ID</p>
```

wurzka commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,608 @@
+      "! Comparison type
+      "! $required
+      comparison_type TYPE zif_aff_types_v1=>ty_object_name_30,
+      "! <p class="shorttext">Check Id</p>
```
```suggestion
      "! <p class="shorttext">Check ID</p>
```

wurzka commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,608 @@
+    "! <p class="shorttext">Inconsistency Category</p>
+    "! Inconsistency category
+    BEGIN OF ty_inconsistency_category,
+      "! <p class="shorttext">Inconsistency Category</p>
+      "! Inconsistency category
+      "! $required
+      result_ui_group  TYPE ty_resultgroup,
```
We suggest having title and name in sync. If you go with `Inconsistency Category`, it might even make sense to change it to `Name`
```suggestion
      "! <p class="shorttext">Name</p>
      "! Inconsistency category
      "! $required
     name  TYPE ty_resultgroup,
```

wurzka commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,608 @@
+      "! Inconsistency category
+      "! $required
+      result_ui_group  TYPE ty_resultgroup,
+      "! <p class="shorttext">Country View Extension</p>
+      "! Country view extension
+      country_xtension TYPE zif_aff_types_v1=>ty_object_name_30,
```
```suggestion
      "! <p class="shorttext">Country View Extension</p>
      "! Country view extension
      country_view_extension TYPE zif_aff_types_v1=>ty_object_name_30,
```

wurzka commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,608 @@
+      "! <p class="shorttext">Result Processes</p>
+      "! Assign result process to the UI group
+      "! $required
+      result_process   TYPE ty_result_processes,
```
```suggestion
      result_processes   TYPE ty_result_processes,
```

wurzka commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,608 @@
+    "! <p class="shorttext">Result Process</p>
+    "! Assign result process to the UI group
+    BEGIN OF ty_results_process,
+      result_process TYPE ty_result_process,
```
```suggestion
    "! <p class="shorttext">Name</p>
      name TYPE ty_result_process,
```

wurzka commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,608 @@
+    "! <p class="shorttext">Check Execution Control</p>
+    "! Check execution control
+    BEGIN OF co_execution_control,
+      "! <p class="shorttext">No Check</p>
+      "! Proceed with check even when inconsistency is identified
+      proceed    TYPE ty_execution_control VALUE '00',
```
Please keep title and name of the component in sync

wurzka commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,608 @@
+    "! <p class="shorttext">Expression Indicator</p>
+    "! Expression indicator
+    BEGIN OF co_expression,
+      "! <p class="shorttext">Begin of Expression</p>
+      "! Begin of expression
+      begin_of TYPE ty_expression_indicator VALUE '/',
+      "! <p class="shorttext">End of Expression</p>
+      "! End of Eexpression
+      end_of   TYPE ty_expression_indicator VALUE '\',
```
```suggestion
      "! <p class="shorttext">Begin of Expression</p>
      "! Begin of expression
      begin_of_expression TYPE ty_expression_indicator VALUE '/',
      "! <p class="shorttext">End of Expression</p>
      "! End of Expression
      end_of_expression   TYPE ty_expression_indicator VALUE '\',
```

KUMARMUKULSAP replied:
Respective node that was using these constants have been removed.

wurzka reviewed (changes requested):
Thanks for updating the AFF. The changes you made look good to me.
Please see my comments.

KUMARMUKULSAP commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,608 @@
+    "! <p class="shorttext">Tax Authority Message Type</p>
+    "! Tax authority message type
+    BEGIN OF ty_message,
+      "! <p class="shorttext">Message Type</p>
+      "! Message type
+      "! $required
+      message_type                TYPE c LENGTH 20,
+      "! <p class="shorttext">Tax Authority Document Type</p>
+      "! Tax authority document type
+      "! $required
+      tax_authority_document_type TYPE c LENGTH 20,
```
@wurzka : Type is important as we configure Tax Authority Document Type as different types of documents, e.g. 'CUSTOMER_INVOICE' / 'SUPPLIER_INVOICE'/ 'DELIVERY_DOCUMENT' etc. Likewise for message type

wurzka replied:
ok, thanks for the explanation. Keeping `type` now makes sense to me.

KUMARMUKULSAP commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,608 @@
+      "! <p class="shorttext">Tax Authority Message Types</p>
+      "! Tax authority message types
+      tax_authority_messages      TYPE ty_messages,
+      "! <p class="shorttext">Additional Tax Authority Tables</p>
+      "! Additional tax authority tables
+      tax_authority_tables        TYPE ty_tax_authority_tables,
```
@wurzka : As part of framework, there is default tax authority table which is not configured in this view. Those configured here are in addition to the default tax authority table which will be used for storing country specific information.

wurzka replied:
ok, I understand. Unfortunately, `additional_tax_authority_tables` is too long, but I believe that the title and name are similar enough here.

KUMARMUKULSAP commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,608 @@
+    "! <p class="shorttext">Representation Type</p>
+    "! Assign representation type associated with consistency scenario
+    BEGIN OF ty_representation_attribute,
+      "! <p class="shorttext">Representation Type</p>
+      "! Representation type
+      "! $required
+      name TYPE zif_aff_types_v1=>ty_object_name_30,
```
Changed

KUMARMUKULSAP replied:
@wurzka , After I did this change, UI looks like this. Can you confirm whether this is ok from your side. I am trying to highlight column description as 'Name' here. 
<img width="1728" alt="Screenshot 2025-04-02 at 4 24 49 PM" src="https://github.com/user-attachments/assets/aa357b06-86c9-407a-8175-bd76f9abd420" />

wurzka replied:
Representation Type and Comparison Type with the field `Name` look good to me.
Regarding Tax Authority Types: Is the `Message Type` a `Tay Authority Message Type`? Then changing the Title of the column (i.e. the component and title of the component) to `Name` would make also sense to me.

KUMARMUKULSAP commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,608 @@
+    "! <p class="shorttext">Additional Tax Authority Table</p>
+    "! Additional tax authority table
+    BEGIN OF ty_tax_authority_table,
+      "! <p class="shorttext">Table Name</p>
+      "! Table name
+      "! $required
+      table_name TYPE zif_aff_types_v1=>ty_object_name_30,
```
Changed

KUMARMUKULSAP commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,608 @@
+      "! Data Source
+      "! $required
+      data_source          TYPE zif_aff_types_v1=>ty_object_name_30,
+      "! <p class="shorttext">Assign Checks</p>
+      "! Assign checks
+      "! $required
+      assign_checks        TYPE ty_assign_checks,
```
Will be changed, once the [question](https://github.com/SAP/abap-file-formats/pull/661#discussion_r2024109724) is cleared

KUMARMUKULSAP replied:
As per the input, I have changed Name and title to _Checks_ and description is changed to 'Assigned checks'

KUMARMUKULSAP commented:
@wurzka : regarding your quesiton https://github.com/SAP/abap-file-formats/pull/661#discussion_r2020701123
If there are more than one check then All checks must have an entry in this node. As it assigns the sequence in which those checks have to be executed. Having an additional node provides better UX to developers in understanding on how the sequence is configured and subsequently execution control assignment would stop what checks if the previous once has error.

wurzka commented:
@KUMARMUKULSAP thank you for the updates.
Regarding the `Check Attributes`, I'd like to involve @schneidermic0 into the discussions.
I understand that for users, it might be easier to configure the execution sequence in an additional table. Do I understand it correctly that the field `execution_sequence` specifies the order of the checks? In this case, one could think about removing the field. Then the order would be specified via the order of the check in the table.
Furthermore, I wondered whether the node should be named something like `Check Execution` instead of `Check Attributes`. More suggestions are welcome :)

KUMARMUKULSAP commented:
> @KUMARMUKULSAP thank you for the updates. Regarding the `Check Attributes`, I'd like to involve @schneidermic0 into the discussions. I understand that for users, it might be easier to configure the execution sequence in an additional table. Do I understand it correctly that the field `execution_sequence` specifies the order of the checks? In this case, one could think about removing the field. Then the order would be specified via the order of the check in the table. Furthermore, I wondered whether the node should be named something like `Check Execution` instead of `Check Attributes`. More suggestions are welcome :)

Let me know if I should schedule a meeting to discuss this to discuss about the idea. Your understanding of execution_sequence is correct. I did not get the suggestion about removing the field and rather having it as entry in the table. Last point about node name can be adjusted as per the suggestion.

schneidermic0 commented:
I had a look at the `check_attributes`. If my understanding is correct, I see two options to improve (and I think Katharina had the same ideas in mind; I just try to rephrase them:

1. We remove `check_attributes` completely. In this case, the fields `execution_sequence` and `execution_control` has to be moved to the check structures like `existence_check` (and maybe further?)

2. We keep the structure `check_attributes` (and, maybe, rename it to something like `check_sequence`. I understood Katharina's question whether we can remove field `execution_sequence` and the order of array/table entries in the list specify the order of execution.

Are the ideas understandable? If so, does any of these options make sense to you to simplify the file format?

KUMARMUKULSAP commented:
> I had a look at the `check_attributes`. If my understanding is correct, I see two options to improve (and I think Katharina had the same ideas in mind; I just try to rephrase them:
> 
> 1. We remove `check_attributes` completely. In this case, the fields `execution_sequence` and `execution_control` has to be moved to the check structures like `existence_check` (and maybe further?)
> 2. We keep the structure `check_attributes` (and, maybe, rename it to something like `check_sequence`. I understood Katharina's question whether we can remove field `execution_sequence` and the order of array/table entries in the list specify the order of execution.
> 
> Are the ideas understandable? If so, does any of these options make sense to you to simplify the file format?

@schneidermic0 @wurzka : I am in favour of 2nd option but without removing the _execution_sequence_. I can rename the node as '_Check Execution Attribute'_. With this change, CHECK EXECUTION SEQUENCE column can also be renamed as EXECUTION SEQUENCE and likewise for CHECK EXECUTION CONTROL as EXECUTION CONTROL.

wurzka commented:
Just one remark (inspired by Michael's proposals): Please try to use inclusive language (see https://help.sap.com/docs/TERMINOLOGY/25cbeaaad3c24eba8ea10b579ce81aa1/83a23df24013403ea4c1fdd0107cc0fd.html -> _Use with caution. Avoid "execute" when you can use "start", "launch", "run", "perform", or "do"._), i.e., if possible, try to avoid `execution`.

KUMARMUKULSAP commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,608 @@
+  TYPES:
+    "! <p class="shorttext">Assign Checks</p>
+    "! Assign checks
+    BEGIN OF ty_assign_checks,
```
changed, ty_assign_checks to ty_checks.

schneidermic0 commented:
>  I am in favour of 2nd option but without removing the execution_sequence.

This is fine for me. I think I would have tried to remove the `execution_sequence` field, but I think this is up to you to keep it.

>  I can rename the node as 'Check Execution Attribute'. With this change, CHECK EXECUTION SEQUENCE column can also be renamed as EXECUTION SEQUENCE and likewise for CHECK EXECUTION CONTROL as EXECUTION CONTROL.

This would be fine for me. For the naming you might want to keep Katharina's comment in mind. 

Here is just an another idea. I tried to gather this with a beginner's mind. It's just an idea. It's no request to change (just food for thoughts): 

- We call the field just `check_sequence` or even `sequence` (we are already in the folder of the checks). Somehow, I still tend to call it `check_sequence`. 
- Within the sequence we define 
   - the `position` to define which position the check takes in the sequence (if you want to keep this) and 
   - `on_failure` with values `proceed_sequence` and `stop_sequence`. 

What do you think?
@wurzka I would be also interested in your opinion on this last proposal

KUMARMUKULSAP commented:
@wurzka @schneidermic0 , all suggested changes are done. Requesting for a re-review.

schneidermic0 commented:
Thanks, Mukul. Our team is on a workshop, today. I am sorry I don't expect we can review it today.

schneidermic0 commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,552 @@
+      "! <p class="shorttext">Always Run Check</p>
+      "! Proceed with check even if inconsistency is identified
+      always_run_check              TYPE ty_control_parameter VALUE '00',
+      "! <p class="shorttext">Run Check if No Inconsistency</p>
+      "! Proceed If no inconsistency identified from previous check
+      run_check_if_no_inconsistency TYPE ty_control_parameter VALUE '01',
```
Minor: I guess I would add "Identified by Previous Check" to the title even if we cannot add it to the field name. But I guess this makes the behavior much more clear.

```suggestion
      "! <p class="shorttext">Run Check if No Inconsistency by Previous Check</p>
      "! Proceed If no inconsistency identified by previous check
      run_check_if_no_inconsistency TYPE ty_control_parameter VALUE '01',
```

KUMARMUKULSAP replied:
If I add by Previous Check in the title then I cannot have variable name that long. I followed the rule to keep title and name in sync

schneidermic0 replied:
You are perfectly right. I really appreciate you are following the rule, in general 👍 

We follow the rule except we can come up with better names in the title that won't fit (due to ABAP's limitation) in the 30 characters of the field name. In this case, having an exception would be fine for me. It is just important, that users can easily map from the title (that might be used in an editor as label or value for enums) to the corresponding field or value in AFF's JSON file. Also this would be the case here, I think.

Maybe, even better would be...? And is it previous check or previous check**s**. You might want to replace "identified" by "found" to shorten the text.

```suggestion
      "! <p class="shorttext">Run Check if No Inconsistency was Identified by Previous Check</p>
      "! Proceed If no inconsistency was identified by previous check
      run_check_if_no_inconsistency TYPE ty_control_parameter VALUE '01',
```

If you like one of my proposal, you can change it. But as mentioned before it's a minor. I would approve this PR without changing it. This could also be changed without incompatible changes. Just let us know your decision.

KUMARMUKULSAP replied:
@schneidermic0 , Thank your for the suggestion. I have made below changes
 ```
"! <p class="shorttext">Always Run Check</p>
      "! Proceed even if inconsistency was found by previous checks
      always_run_check              TYPE ty_control_parameter VALUE '00',
      "! <p class="shorttext">Run Check if No Inconsistency was found by Previous Checks</p>
      "! Proceed If no inconsistency was found by previous checks
      run_check_if_no_inconsistency TYPE ty_control_parameter VALUE '01',
```

schneidermic0 replied:
looks good to me. I will close this discussion as soon you have pushed the changes... Thanks. 👍

schneidermic0 replied:
Just as a heads up. I don't see this improvement in the latest version.

schneidermic0 commented on file-formats/edcc/examples/z_aff_example_edcc.edcc.json:
```diff
@@ -0,0 +1,77 @@
+      "name": "TAXAUTHDOC"
+    }
+  ],
+  "additionalSelectionFields": [],
+  "taxAuthorityMessageTypes": [],
+  "taxAuthorityTables": [],
+  "edocumentTypes": [],
```
Is the example really serialized this way by AFF? Since the tables are not marked as `$required`, I would have expected they are not generated if they are empty.

```suggestion
```

wurzka replied:
Did you copy the example from our internal tooling? This does not serialize the JSON correctly (wrt non requried empty tables).
Regarding your question on validation, I suggest moving this discussion to the other channel. Here, it should be enough removing the empty tables.

schneidermic0 replied:
I suggest to just remove the empty tables here in the repository. And we follow-up for our export tool whether we need to fix anything here separately. Is this fine for you?

KUMARMUKULSAP replied:
@wurzka: example is created using the ADT Form UI Dev Tool. Is this the one that you are referring to as internal tool?

schneidermic0 commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,552 @@
+  "! <p class="shorttext">eDocument Types</p>
+  "! Assign eDocument types associated with consistency scenario
+  TYPES ty_edoc_types  TYPE SORTED TABLE OF ty_edoc_type WITH UNIQUE KEY name.
+  "! <p class="shorttext">Inconsistency Category</p>
+  "! Inconsistency category
+  TYPES ty_resultgroup TYPE c LENGTH 20.
```
Minor: Do you want to rename this to `ty_inconsistency_category`?

KUMARMUKULSAP replied:
There is already a `ty_inconsistency_category` declared at line 390

schneidermic0 commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,552 @@
+      "! <p class="shorttext">Content Checks</p>
+      "! Assigned content checks
+      content_checks  TYPE ty_content_checks,
+      "! <p class="shorttext">Run Attributes</p>
+      "! Check run attributes
+      run_attributes  TYPE ty_run_attributes,
```
Minor: Just as information. You can just resolve this comment. 

I preferred my suggestion with `sequence` or `check_sequence`. I guess for the user it becomes more clear instead of "run attributes".

However, this is your decision. I just want to share my feedback. :)

schneidermic0 commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,552 @@
+      "! Check id
+      "! $required
+      check_id          TYPE zif_aff_types_v1=>ty_object_name_30,
+      "! <p class="shorttext">Sequence</p>
+      "! Check run sequence
+      "! $required
+      run_sequence      TYPE ty_run_sequence,
```
Minor: Mart this as resolved, too. Again, only feedback:

For me this is not the sequence, but the position within the sequence

schneidermic0 commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,552 @@
+      "! Check run sequence
+      "! $required
+      run_sequence      TYPE ty_run_sequence,
+      "! <p class="shorttext">Control Parameter</p>
+      "! Check execution control parameter
+      control_parameter TYPE ty_control_parameter,
```
Minor: Mark this as resolved, too. Again, only feedback:

Do you plan other settings than how the check sequence shall behave if the check raises an error. If not, I think it would make it more clear to specify something like "behavior on failure". I like the new possible values for this field: "Always Run Check" and "Run Check if No Inconsistency" (or "Run Check if No Inconsistency by Previous Check"). They are quite clear to me.

schneidermic0 reviewed (changes requested):
Beside a comment from last November regarding "General Information" that seems not to be addressed, yet. Otherwise I found only minor changes or wanted to add some feedback.

Therefore, if Katharina does not find anything I missed, I think we can take-up this version also for the UX cross check, if you don't want to incorporate my minor feedback.

wurzka reviewed (commented):
Besides the open issues raised in @schneidermic0's comments, AFF looks really good to me. Thanks for the cooperation @KUMARMUKULSAP!

KUMARMUKULSAP commented:
@schneidermic0 @wurzka : Please review this commit for AFF with review comments fix.

wurzka commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,545 @@
+      "! <p class="shorttext">Content Checks</p>
+      "! Assigned content checks
+      content_checks  TYPE ty_content_checks,
+      "! <p class="shorttext">Checks Sequence</p>
+      "! Checks sequence and control parameter
+      checks_sequence TYPE ty_checks_sequence,
```
I'm struggling with the naming here. 
```suggestion
      "! <p class="shorttext">Check Sequence</p>
      "! Sequence of the checks and control parameter
      check_sequence TYPE ty_checks_sequence,
```
`Check Sequence` sounds better to me. @KUMARMUKULSAP , @schneidermic0 , what do you think?

KUMARMUKULSAP replied:
@wurzka : I chose to put Checks as it will specify sequence of all assigned checks. i.e. all checks those are assigned must have an entry if and only if there are multiple checks.

schneidermic0 replied:
I agree with Katharina. `checks_sequence` sounds strange to me. I have also double checked with our UA. A sequence indicates that several checks are involved, doesn't it? Another option would be "Sequence of Checks".

wurzka reviewed (commented):
Just one minor. Else looks good to me.

schneidermic0 reviewed (commented):
@KUMARMUKULSAP Thanks for all the updates.

Accept the "check sequence" (I agree to Katharina's comment), it looks good to me.  

Just as a heads-up: 
1. I found two small regressions in the example 
2. one addition you wanted to add seems to be missing

KUMARMUKULSAP commented:
@schneidermic0 : Regression issues are manually fixed. And missed correction is now in place

schneidermic0 commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -529,10 +529,10 @@ INTERFACE zif_aff_edcc_v1
       always_run_check              TYPE ty_control_parameter VALUE '00',
       "! <p class="shorttext">Run Check if No Inconsistency</p>
-      "! Proceed If no inconsistency identified from previous check
+      "!Proceed If no inconsistency was found by previous checks
```
Thanks for adapting the description. However, the title/shorttext is what users will see in UI. Maybe, you want to adapt the shortest, too.

See also your own proposal here: ;)
https://github.com/SAP/abap-file-formats/pull/661#discussion_r2028471259

schneidermic0 replied:
I guess there is a misunderstanding. Above you mentioned following code:

```abap
     "! <p class="shorttext">Always Run Check</p>
     "! Proceed even if inconsistency was found by previous checks
     always_run_check              TYPE ty_control_parameter VALUE '00',
     "! <p class="shorttext">Run Check if No Inconsistency was found by Previous Checks</p>
     "! Proceed If no inconsistency was found by previous checks
     run_check_if_no_inconsistency TYPE ty_control_parameter VALUE '01',
```

For me it was helpful to have the information that previous checks don't found any inconsistency. As by you in following line:

```abap
     "! Proceed If no inconsistency was found by previous checks
```

In the current AFF, I see only:

```abap
"! <p class="shorttext">Run Check if No Inconsistency</p>
```

schneidermic0 reviewed (approved):
Thank you for all of your adaptations, @kumamar.

It looks really good to me, now. I approve this PR.

Just see my comment and tell me whether you want to adapt it before we merge this change.

albertmink commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,552 @@
+      "! <p class="shorttext">Country</p>
+      "! Country
+      "! $required
+      country TYPE c LENGTH 3,
```
Looking at the UI/editor, I assume this should be `char2`

albertmink replied:
type is changed to char2

KUMARMUKULSAP commented:
@schneidermic0 @albertmink : Review comments from UX cross check 2 as well as below changes have been committed 
    ` "! <p class="shorttext">Run Check if No Inconsistency was found by Previous Checks</p>`
Please review and provide your sign off.

albertmink commented on file-formats/edcc/type/zif_aff_edcc_v1.intf.abap:
```diff
@@ -0,0 +1,552 @@
+      "! <p class="shorttext">Country Code</p>
+      "! Country code
+      "! $required
+      country TYPE c LENGTH 2,
```
It's best practise to align names and titles. 
```suggestion
      countryCode TYPE c LENGTH 2,
```

schneidermic0 reviewed (approved):
Thanks, @KUMARMUKULSAP. Looks good to me.

albertmink reviewed (approved):
LGTM

wurzka reviewed (approved):
Thanks, looks good t me.
